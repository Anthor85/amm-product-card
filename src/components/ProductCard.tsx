import React, { createContext, JSX } from 'react';
import { useProduct } from '../hooks/useProduct';
import {
  InitialValues,
  onChangeArgs,
  Product,
  ProductCardHandlers,
  ProductContextProps,
} from '../interfaces/interfaces';

import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  product: Product;
  value?: number;
  onChange?: (args: onChangeArgs) => void;
  // children?: ReactElement | ReactElement[];
  children: (props: ProductCardHandlers) => JSX.Element;
  className?: string;
  style?: React.CSSProperties;
  initialValues?: InitialValues;
}

export const ProductCard = ({
  children,
  product,
  value,
  onChange,
  className = '',
  style,
  initialValues,
}: Props) => {
  const {
    counter,
    increaseBy,
    isMaxCountReached,
    maxCount,
    reset,
  } = useProduct({
    product,
    value,
    initialValues,
    onChange,
  });

  return (
    <Provider value={{ counter, increaseBy, product, maxCount }}>
      <div
        className={`${styles.productCard} ${className}`}
        key={product.id}
        style={style}
      >
        {children({
          count: counter,
          isMaxCountReached,
          maxCount,
          product,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};
