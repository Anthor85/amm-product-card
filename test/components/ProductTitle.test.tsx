import React from 'react';
import render from 'react-test-renderer';
import { ProductCard, ProductTitle } from '../../src/components';
import { product1 } from '../data/products';

describe('ProductTitle', () => {
  test('debe mostrar e componente con el título personalizado', () => {
    const wrapper = render.create(
      <ProductTitle
        title="Título personalizado"
        className="titulo-personalizado"
      />
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('debe mostrar el componente con el nombre del producto', () => {
    const wrapper = render.create(
      <ProductCard product={product1}>{() => <ProductTitle />}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
