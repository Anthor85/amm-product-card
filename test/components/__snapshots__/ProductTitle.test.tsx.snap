// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle debe mostrar e componente con el título personalizado 1`] = `
<span
  className="productDescription titulo-personalizado"
>
  Título personalizado
</span>
`;

exports[`ProductTitle debe mostrar el componente con el nombre del producto 1`] = `
<div
  className="productCard "
>
  <span
    className="productDescription "
  >
    Coffee Mug - Sin Imagen
  </span>
</div>
`;
